### PR TITLE
Add bastion host controller

### LINKED ISSUE #165: Add bastion host controller
# Summary

Add a bastion host controller analogous to the GCP provider here: https://github.com/gardener/gardener-extension-provider-gcp/tree/master/pkg/controller/bastion

The bastion host contract in the Gardener extension can be found here: https://github.com/gardener/gardener/blob/master/docs/extensions/bastion.md

The Gardener proposal describing the bastion host integration can be found here: https://github.com/gardener/gardener/blob/master/docs/proposals/15-manage-bastions-and-ssh-key-pair-rotation.md

Here we need to implement the following:
- Implement the bastion controller
   - Create `Machine` and `NetworkInterface`
   - Create a `VirtualIP`
   - Create `NetworkPolicies` restricting the traffic to the bastion host 
- Register the bastion host controller 

A high level architecture of the target picture will look like the following:
Screenshot 2023-04-19 at 10 10 40

## Definition of done

- Add controller coding to create and manage a bastion host
- Add suite tests ensuring the correct behaviour of the controller

### PR DESCRIPTION
**Pull request by sujeet01**
This adds support for the new `Bastion` resource, which is part of [GEP-15](https://github.com/gardener/gardener/blob/master/docs/proposals/15-manage-bastions-and-ssh-key-pair-rotation.md)'s new SSH handling. A new controller takes care of creating a bastion host with the configured SSH key and allowing ingress from that host to the other worker nodes in a shoot.

- Add `BastionConfig` type
- Register bastion controller with manager
- Add bastion host reconciler and tests
- Add bastion host delete and tests
- Add bastion config validator and tests
- Add suite tests

Resolves #165

### CODE REVIEW

File: pkg/controller/bastion/actuator_reconcile.go [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+}
+
+// Reconcile implements bastion.Actuator.
+func (a *actuator) Reconcile(ctx context.Context, log logr.Logger, bastion *extensionsv1alpha1.Bastion, cluster *controller.Cluster) error {
+	return a.reconcile(ctx, log, bastion, cluster)
+}
+
+func (a *actuator) reconcile(ctx context.Context, log logr.Logger, bastion *extensionsv1alpha1.Bastion, cluster *controller.Cluster) error {
+	log.V(2).Info("Reconciling bastion host")
+	err := bastionConfigCheck(a.bastionConfig)
```
Comment (afritzler) [inline pkg/controller/bastion/actuator_reconcile.go:61]:
```suggestion
	if err := bastionConfigCheck(a.bastionConfig); err != nil {
```

File: pkg/controller/bastion/actuator_reconcile.go [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+	if err != nil {
+		return err
```
Comment (afritzler) [inline pkg/controller/bastion/actuator_reconcile.go:63]: Please wrap the returned error with a proper message.

File: pkg/controller/bastion/actuator_reconcile.go [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+			// requeue rather soon, so that the user (most likely gardenctl eventually)
+			// doesn't have to wait too long for the public endpoint to become available
+			RequeueAfter: 5 * time.Second,
+			Cause:        fmt.Errorf("bastion instance has no public/private endpoints yet"),
+		}
+	}
+
+	// once a public endpoint is available, publish the endpoint on the
+	// Bastion resource to notify upstream about the ready instance
+	log.V(2).Info("Successfully reconciled bastion host")
```
Comment (afritzler) [inline pkg/controller/bastion/actuator_reconcile.go:110]:
```suggestion
	log.V(2).Info("Reconciled bastion host")
```

File: pkg/controller/bastion/actuator_reconcile.go [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+	patch := client.MergeFrom(bastion.DeepCopy())
+	bastion.Status.Ingress = endpoints.public
```
Comment (afritzler) [inline pkg/controller/bastion/actuator_reconcile.go:112]: What about the private endoint?

File: pkg/controller/bastion/actuator_reconcile.go [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+
+	if len(machine.Status.NetworkInterfaces) == 0 {
+		return nil, fmt.Errorf("no network interface found for machine: %s", machine.Name)
+	}
+
+	if len(machine.Status.NetworkInterfaces[0].IPs) == 0 {
+		return nil, fmt.Errorf("no private ip found for network interface for machine: %s", machine.Name)
+	}
+
+	privateIP := machine.Status.NetworkInterfaces[0].IPs[0].String()
```
Comment (afritzler) [inline pkg/controller/bastion/actuator_reconcile.go:139]: We need to make sure that we return here the IPv4 address of the network interface as we can have multiple addresses here. Also please add a TODO here to handle the IPv6 case once the Gardener extension has proper support for IPv6.

File: pkg/controller/bastion/actuator_reconcile.go [hunk 6]
```diff
@@ -0,0 +1,327 @@
...
+	virtualIP := machine.Status.NetworkInterfaces[0].VirtualIP.String()
+	if ingress := addressToIngress(&machine.Name, &virtualIP); ingress != nil {
+		endpoints.public = ingress
+	}
+
+	return endpoints, nil
+}
+
+// bastionConfigCheck checks whether a bastion configuration is valid.
+func bastionConfigCheck(bastionConfig *controllerconfig.BastionConfig) error {
```
Comment (afritzler) [inline pkg/controller/bastion/actuator_reconcile.go:157]: Please rename this method to `validateConfiguration`.

Comment (afritzler) [inline pkg/controller/bastion/actuator_reconcile.go:157]: Just call the variable `config` as we know that we have a bastion host configuration here.

File: pkg/controller/bastion/actuator_reconcile.go [hunk 7]
```diff
@@ -0,0 +1,327 @@
...
+		return fmt.Errorf("bastion not supported as no Image is configured for the bastion host machine")
+	}
+	return nil
+}
+
+// applyMachine applies the configuration to create or update the bastion host machine
+// and the ignition secret used for provisioning the machine. It first sets the owner
+// reference for the ignition secret to the bastion host machine, to ensure that the secret
+// is garbage collected when the machine is deleted.
+func applyMachine(ctx context.Context, onmetalClient client.Client, ignitionSecret *corev1.Secret, bastionHost *computev1alpha1.Machine) (*computev1alpha1.Machine, error) {
```
Comment (afritzler) [inline pkg/controller/bastion/actuator_reconcile.go:176]: Please rename this to `applyMachineAndIgnitionSecret`.

File: pkg/controller/bastion/actuator_reconcile.go [hunk 8]
```diff
@@ -0,0 +1,327 @@
...
+
+	_, err := controllerutil.CreateOrPatch(ctx, onmetalClient, bastionHost, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create or patch bastion host machine %s: %w", client.ObjectKeyFromObject(bastionHost), err)
+	}
+
+	//TODO: Even after its owner is set as the Bastion machine, ignitionSecret does not get garbage collected when deleting Bastion.
```
Comment (afritzler) [inline pkg/controller/bastion/actuator_reconcile.go:183]: Does this issue still occur?

File: pkg/controller/bastion/actuator_reconcile.go [hunk 9]
```diff
@@ -0,0 +1,327 @@
...
+	_, err = controllerutil.CreateOrPatch(ctx, onmetalClient, ignitionSecret, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create or patch ignition secret %s for bastion host %s: %w", client.ObjectKeyFromObject(ignitionSecret), client.ObjectKeyFromObject(bastionHost), err)
+	}
+
+	return bastionHost, nil
+}
+
+// createIgnitionSecret constructs a Kubernetes secret object containing an ignition file for the Bastion machine
+func createIgnitionSecret(namespace string, opt *Options) (*corev1.Secret, error) {
```
Comment (afritzler) [inline pkg/controller/bastion/actuator_reconcile.go:197]: Lets call this one `generateIgnitionSecret`.

File: pkg/controller/bastion/actuator_reconcile.go [hunk 10]
```diff
@@ -0,0 +1,327 @@
...
+		return nil, fmt.Errorf("failed to create ignition file for machine %s: %w", opt.BastionInstanceName, err)
+	}
+
+	ignitionData := map[string][]byte{}
+	ignitionData[computev1alpha1.DefaultIgnitionKey] = []byte(ignitionContent)
+	ignitionSecret := &corev1.Secret{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: corev1.SchemeGroupVersion.String(),
+			Kind:       "Secret",
+		},
```
Comment (afritzler) [inline pkg/controller/bastion/actuator_reconcile.go:216]: `TypeMeta` should not be necessary here.

File: pkg/controller/bastion/actuator_reconcile.go [hunk 11]
```diff
@@ -0,0 +1,327 @@
...
+			Namespace: namespace,
+		},
+		Data: ignitionData,
+	}
+
+	return ignitionSecret, nil
+}
+
+// createMachine constructs a Machine object for the Bastion instance
+func createMachine(namespace string, bastionConfig *controllerconfig.BastionConfig, infraStatus *api.InfrastructureStatus, BastionInstanceName string, ignitionSecretName string) *computev1alpha1.Machine {
```
Comment (afritzler) [inline pkg/controller/bastion/actuator_reconcile.go:228]: Let's call this one `generateMachine`.

File: pkg/controller/bastion/actuator_reconcile.go [hunk 12]
```diff
@@ -0,0 +1,327 @@
...
+	}
+
+	// TODO: Add NetworkPolicy related implementation
+
+	ignitionSecret, err := createIgnitionSecret(namespace, opt)
+	if err != nil {
+		return fmt.Errorf("failed to create ignition secret: %w", err)
+	}
+
+	bastionMachine := createMachine(namespace, a.bastionConfig, infraStatus, opt.BastionInstanceName, ignitionSecret.Name)
```
Comment (afritzler) [inline pkg/controller/bastion/actuator_reconcile.go:87]: Can you move this into the `applyMachineAndIngitionSecret` method.

File: pkg/controller/bastion/actuator_reconcile_test.go [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+		Eventually(Object(bastionMachine)).Should(SatisfyAll(
+			HaveField("Spec.MachineClassRef", corev1.LocalObjectReference{
+				Name: "my-machine-class",
+			}),
+			HaveField("Spec.Image", "my-image"),
+			HaveField("Spec.IgnitionRef.Name", getIgnitionNameForMachine(bastionMachine.Name)),
+			HaveField("Spec.Power", computev1alpha1.PowerOn),
+			HaveField("Spec.NetworkInterfaces", ContainElement(SatisfyAll(
+				HaveField("Name", "primary"),
+				HaveField("NetworkInterfaceSource.Ephemeral.NetworkInterfaceTemplate.Spec.NetworkRef.Name", "my-network"),
```
Comment (afritzler) [inline pkg/controller/bastion/actuator_reconcile_test.go:96]: How about the other network interface fields?

File: pkg/controller/infrastructure/actuator_reconcile_test.go [hunk 14]
```diff
@@ -102,111 +103,6 @@ var _ = Describe("Infrastructure Reconcile", func() {
 			},
 		}
 
-		Eventually(Object(natGateway)).Should(SatisfyAll(
```
Comment (afritzler) [inline pkg/controller/infrastructure/actuator_reconcile_test.go:105]: Why has this code been removed?

File: pkg/controller/bastion/configvalidator.go [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+		return fmt.Errorf("bastion spec userdata can't be empty")
+	}
+
+	for _, ingress := range bastion.Spec.Ingress {
+		if ingress.IPBlock.CIDR == "" {
+			return fmt.Errorf("bastion spec Ingress CIDR can't be empty")
+		}
+		_, _, err := net.ParseCIDR(ingress.IPBlock.CIDR)
+		if err != nil {
+			return fmt.Errorf("invalid bastion spec Ingress CIDR %w", err)
```
Comment (afritzler) [inline pkg/controller/bastion/configvalidator.go:95]:
```suggestion
			return fmt.Errorf("invalid bastion spec Ingress CIDR: %w", err)
```

File: pkg/controller/bastion/configvalidator.go [hunk 16]
```diff
@@ -0,0 +1,146 @@
...
+		allErrs = append(allErrs, field.InternalError(nil, err))
+		return allErrs
+	}
+
+	return allErrs
+}
+
+func validateBastion(bastion *extensionsv1alpha1.Bastion) error {
+	if bastion == nil {
+		return fmt.Errorf("bastion can't be nil")
```
Comment (afritzler) [inline pkg/controller/bastion/configvalidator.go:82]:
```suggestion
		return fmt.Errorf("bastion can not be nil")
```

Comment (afritzler) [reply pkg/controller/bastion/configvalidator.go:82]: Please use "can not" instead of "can't".

**Review by afritzler — CHANGES_REQUESTED** [review]

**Comment by afritzler** [comment]
We also need to add the corresponding helm chart values similar to this example: https://github.com/gardener/gardener-extension-provider-openstack/blob/master/charts/gardener-extension-provider-openstack/values.yaml#L86 and https://github.com/gardener/gardener-extension-provider-openstack/blob/master/charts/gardener-extension-provider-openstack/templates/configmap.yaml#LL28C1-L30C62

File: pkg/controller/bastion/actuator_reconcile.go (hunk 8 above)
Comment (sujeet01) [reply pkg/controller/bastion/actuator_reconcile.go]: yes, able to set owner reference correctly, but still ignition secret is not getting garbage collected so we have to explicitly remove it.

File: pkg/controller/bastion/actuator_delete.go [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+			Namespace: namespace,
+		},
+	}
+
+	if err := onmetalClient.Delete(ctx, bastionMachine); err != nil {
+		return fmt.Errorf("error deleting bastion machine: %v", err)
+	}
+
+	if err := onmetalClient.Delete(ctx, ignitionSecret); err != nil {
+		return fmt.Errorf("error deleting ignition secret: %v", err)
```
[the remark this answers is not in the document]
Comment (sujeet01) [reply pkg/controller/bastion/actuator_delete.go:72]: Even though owner reference is set correctly, still ignition secret is not getting garbage collected so we have to explicitly remove it.

File: pkg/controller/bastion/actuator_delete.go [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+			Namespace: namespace,
+			Name:      bastionMachineName,
+		},
+	}
+	ignitionSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      getIgnitionNameForMachine(bastionMachineName),
+			Namespace: namespace,
+		},
+	}
```
[the remark this answers is not in the document]
Comment (sujeet01) [reply pkg/controller/bastion/actuator_delete.go:65]: Even though owner reference is set correctly, still ignition secret is not getting garbage collected so we have to explicitly remove it.

File: pkg/controller/bastion/actuator_reconcile.go (hunk 5 above)
Comment (sujeet01) [reply pkg/controller/bastion/actuator_reconcile.go]: Done.

File: pkg/controller/bastion/actuator_reconcile.go (hunk 4 above)
Comment (sujeet01) [reply pkg/controller/bastion/actuator_reconcile.go:105]: Doesn't that come later, when we add network policy implementation?

File: pkg/controller/bastion/actuator_delete.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (afritzler) [reply pkg/controller/bastion/actuator_delete.go]: The reason for that is, that the GC in the `kube-controller-manager` is not running in the `envtest` environment. In your test you just need to ensure that the `ownerRef` is set correctly. The rest should be handled by k8s. There is also no need to test the GC capabilities as this is tested by the community already.

File: pkg/controller/infrastructure/actuator_reconcile_test.go (hunk 14 above)
Comment (kasabe28) [reply pkg/controller/infrastructure/actuator_reconcile_test.go]: I thought testcase was duplicated, but I was wrong. Reverting changes.

**Review by afritzler — CHANGES_REQUESTED** [review]

File: pkg/controller/bastion/utils.go [hunk 19]
```diff
@@ -0,0 +1,95 @@
...
+	bastionName := bastion.Name
+	if bastionName == "" {
+		return "", fmt.Errorf("bastionName can not be empty")
+	}
+	if clusterName == "" {
+		return "", fmt.Errorf("clusterName can not be empty")
+	}
+	staticName := clusterName + "-" + bastionName
+	nameSuffix := strings.Split(string(bastion.UID), "-")[0]
+	return fmt.Sprintf("%s-bastion-%s", staticName, nameSuffix), nil
```
[the remark this answers is not in the document]
Comment (sujeet01) [reply pkg/controller/bastion/utils.go]: Added length check.

**Review by afritzler — APPROVED** [review]

**Pull request merged**